rust/src/web/graphql/transactions/mod.rs

```diff
@@ -0,0 +1,303 @@
+use super::{date_time_to_scalar, db, F64Ord};
+use crate::{
+    block::BlockHash,
+    canonicity::{store::CanonicityStore, Canonicity},
+    command::signed::{SignedCommand, SignedCommandWithData},
+    ledger::public_key::PublicKey,
+    protocol::serialization_types::staged_ledger_diff::{
+        SignedCommandPayloadBody, StakeDelegation,
+    },
+    store::{
+        user_commands_iterator, user_commands_iterator_signed_command,
+        user_commands_iterator_txn_hash,
+    },
+    web::graphql::{gen::TransactionQueryInput, DateTime},
+};
+use async_graphql::{Context, Enum, Object, Result, SimpleObject};
+use std::cmp::Ordering;
+
+#[derive(Default)]
+pub struct TransactionsQueryRoot;
+
+const NANO_F64: f64 = 1_000_000_000_f64;
+
+#[Object]
+impl TransactionsQueryRoot {
+    pub async fn transaction(
+        &self,
+        _ctx: &Context<'_>,
+        _query: TransactionQueryInput,
+    ) -> Result<Option<Transaction>> {
+        todo!()
+    }
+
+    pub async fn transactions(
+        &self,
+        ctx: &Context<'_>,
+        query: TransactionQueryInput,
+        limit: Option<i64>,
+        sort_by: TransactionSortByInput,
+    ) -> Result<Vec<Option<Transaction>>> {
+        let db = db(ctx);
+        let limit = limit.unwrap_or(100);
+        let limit_idx = limit as usize;
+
+        let mut transactions: Vec<Option<Transaction>> = Vec::new();
+
+        let iter = user_commands_iterator(db);
+
+        for entry in iter {
+            let txn_hash = user_commands_iterator_txn_hash(&entry)?;
+
+            if let Some(query_txn_hash) = query.hash.to_owned() {
+                if txn_hash != query_txn_hash {
+                    continue;
+                }
+            }
+
+            let cmd = user_commands_iterator_signed_command(&entry)?;
+
+            let block_state_hash = cmd.state_hash.to_owned();
+            let block_date_time = date_time_to_scalar(cmd.date_time as i64);
+
+            let canonical = match db.get_block_canonicity(&block_state_hash.to_owned())? {
+                Some(canonicity) => canonicity == Canonicity::Canonical,
+                None => false,
+            };
+
+            let transaction =
+                Transaction::from_cmd(cmd, block_date_time, &block_state_hash, canonical);
+
+            // Only add transactions that satisfy the input query
+            if query.matches(&transaction) {
+                transactions.push(Some(transaction));
+            };
+
+            // Early break if the transactions reach the query limit
+            if transactions.len() >= limit_idx {
+                break;
+            }
```

Comment:
The txns would have had to be in sorted order already for this early exit to be correct I think. E.g. Limit 1 and sort by NonceDesc. There would be no guarantee that the largest nonce would be returned. Unfortunately you'll need to look at an index first or scan through all the transactions

Reply:
Excellent catch. I'll fix that.

Reply:
[Fixed](https://github.com/Granola-Team/mina-indexer/pull/769/files#diff-22e46ac9e2b2d33676e8ff2a5489995088666459acf57a10bfa88b6ea6c5dac1R77) by removing `if` statement and using `truncate`.